File: app/src/main/java/ai/elimu/content_provider/util/MultimediaDownloader.java
package ai.elimu.content_provider.util;

import android.util.Log;

import org.apache.commons.io.IOUtils;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;

public class MultimediaDownloader {

    public static byte[] downloadFileBytes(String urlValue) {
        Log.i(MultimediaDownloader.class.getName(), "downloadMultimedia");

        Log.i(MultimediaDownloader.class.getName(), "Downloading from " + urlValue);

        byte[] bytes = null;

        try {
            URL url = new URL(urlValue);

            HttpURLConnection httpURLConnection = (HttpURLConnection) url.openConnection();
            httpURLConnection.setRequestMethod("GET");
            httpURLConnection.connect();

            int responseCode = httpURLConnection.getResponseCode();
            Log.i(MultimediaDownloader.class.getName(), "responseCode: " + responseCode);
            InputStream inputStream = null;
            if (responseCode == 200) {
                inputStream = httpURLConnection.getInputStream();
                bytes = IOUtils.toByteArray(inputStream);
            } else {
                inputStream = httpURLConnection.getErrorStream();
                BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(inputStream));
                String response = "";
                String line;
                while ((line = bufferedReader.readLine()) != null) {
                    response += line;
                }
                Log.e(MultimediaDownloader.class.getName(), "responseCode: " + responseCode + ", response: " + response);
            }
        } catch (IOException e) {
            Log.e(MultimediaDownloader.class.getName(), null, e);
        }

        return bytes;
    }
}

File: app/src/main/java/ai/elimu/content_provider/util/MultimediaDownloader.kt
package ai.elimu.content_provider.util

import android.util.Log
import org.apache.commons.io.IOUtils
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL

object MultimediaDownloader {
    
    private const val TAG = "MultimediaDownloader"
    
    fun downloadFileBytes(urlValue: String): ByteArray? {
        Log.i(TAG, "downloadFileBytes")

        Log.i(TAG, "Downloading from $urlValue")

        var bytes: ByteArray? = null

        try {
            val url = URL(urlValue)

            val httpURLConnection = url.openConnection() as HttpURLConnection
            httpURLConnection.requestMethod = "GET"
            httpURLConnection.connect()

            val responseCode = httpURLConnection.responseCode
            Log.i(
                TAG,
                "responseCode: $responseCode"
            )
            val inputStream: InputStream?
            if (responseCode == 200) {
                inputStream = httpURLConnection.inputStream
                bytes = IOUtils.toByteArray(inputStream)
            } else {
                inputStream = httpURLConnection.errorStream
                val bufferedReader = BufferedReader(InputStreamReader(inputStream))
                var response = ""
                var line: String
                while ((bufferedReader.readLine().also { line = it }) != null) {
                    response += line
                }
                Log.e(
                    TAG,
                    "responseCode: $responseCode, response: $response"
                )
            }
        } catch (e: IOException) {
            Log.e(TAG, null, e)
        }

        return bytes
    }
}
